modules/gdscript/gdscript_abstract_codegen.h

```diff
@@ -0,0 +1,165 @@
+/* Copyright (c) 2007-2020 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2020 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef GDSCRIPT_ABSTRACT_CODEGEN
+#define GDSCRIPT_ABSTRACT_CODEGEN
+
+#include "core/io/multiplayer_api.h"
+#include "core/string_name.h"
+#include "core/variant.h"
+#include "gdscript_function.h"
+#include "gdscript_functions.h"
+
+class GDScriptAbstractCodeGenerator {
+public:
+	struct Address {
+		enum AddressMode {
+			SELF,
+			CLASS,
+			MEMBER,
+			CONSTANT,
+			CLASS_CONSTANT,
+			LOCAL_CONSTANT,
+			LOCAL_VARIABLE,
+			FUNCTION_PARAMETER,
+			TEMPORARY,
+			GLOBAL,
+			NAMED_GLOBAL,
+			NIL,
+		};
+		AddressMode mode = NIL;
+		uint32_t address = 0;
+		StringName name; // For named variables.
+		GDScriptDataType type;
+
+		Address() {}
+		Address(AddressMode p_mode, const GDScriptDataType &p_type = GDScriptDataType()) {
+			mode = p_mode;
+			type = p_type;
+		}
+		Address(AddressMode p_mode, uint32_t p_address, const GDScriptDataType &p_type = GDScriptDataType()) {
+			mode = p_mode,
+			address = p_address;
+			type = p_type;
+		}
+		Address(AddressMode p_mode, const StringName &p_name, const GDScriptDataType &p_type = GDScriptDataType()) {
+			mode = p_mode,
+			name = p_name;
+			type = p_type;
+		}
+	};
+
+	virtual uint32_t add_parameter(const StringName &p_name, bool p_is_optional, const GDScriptDataType &p_type) = 0;
```

Comment:
I am not very keen on using either uint32_t or the address struct. If you ask me, I would just get rid of the Address struct and just use something like:

```typedef uint32_t Address; ```

everywhere for readability. I also think feeding the address struct as parameter does not make much sense because you already have the information such as the name already provided when you registered.